src/Codegen/TypeDef.hs

```diff
@@ -1,10 +1,87 @@
+{-# LANGUAGE QuasiQuotes #-}
 module Codegen.TypeDef where
 
 import qualified IR.Types.Flat                 as L
 import qualified IR.Types.TypeSystem           as L
 
+import           Language.C.Quote.GCC
 import qualified Language.C.Syntax             as C
 
+import           Common.Identifiers             ( DConId
+                                                , TConId
+                                                , fromString
+                                                , ident
+                                                )
+
+import           Codegen.Identifiers
+import           Data.Char                      ( toLower )
+import           Data.List                      ( maximumBy
+                                                , partition
+                                                )
+
 -- | Generate definitions for SSM type definitions.
-genTypeDef :: L.TypeDef L.Type -> C.Definition
-genTypeDef = error "Not yet implemented"
+genTypeDef :: TConId -> L.TypeDef L.Type -> (C.Definition, C.Definition)
+genTypeDef tconid (L.TypeDef dCons _) = (tags, structDef)
+ where
+  structDef = [cedecl| typedef struct {
+                       $ty:ssm_mm_md* $id:header;
```

Comment:
Both the header field and the payload field should be values rather than pointers.

Reply:
Ah, yes, I will fix that header to be a value! 
I have a question about the payload being a value - I tried to do this, but the C compiler won't allow it because of cyclic association. In our meeting last week we came up with a struct definition like this:
```
struct ssm_object_t{               // a generic object
	struct ssm_mm_md header;       // gc header
	struct ssm_object_t payload[1];  // causes compiler error! 
};
```
We get a compiler error here because ssm_object_t cannot hold an array of itself/size of the array would be unknown. Because of this error, I changed payload to an array of ssm_object_t pointers like this:
```
struct ssm_object_t{               // a generic object
	struct ssm_mm_md header;       // gc header
	struct ssm_object_t* payload[1];  // compiles okay
};
```
Is there a better way to solve this issue? I think the root of this issue is the initialization code we came up with at the meeting, which never actually use our adt-specific struct, only ssm_object_t:
```
void main(){
	// (max size across all data constructors, tag)
	struct ssm_object_t* a = ssm_new(2, MyFalseTag);
	a->payload[0] = (struct ssm_object_t*) ssm_int_obj(42);
	a->payload[1] = (struct ssm_object_t*) ssm_int_obj(18);
}
```
Is the better way to solve this cyclic association problem to use the adt-specific struct somehow instead of the ssm_object_t struct in this initialization code? Maybe like this??
```
struct MyBool{
	struct ssm_mm_md header;
	struct ssm_object_t payload[2]; //max number of fields	
};
void main(){
	// (max size across all data constructors, tag)
	struct MyBool a = ssm_new(2, MyFalseTag);
	a.payload[0] = ssm_int_obj(42);
	a.payload[1] = ssm_int_obj(18);
}
```

Reply:
There are two types: the single-word value or pointer:

```
union ssm_object_t {
  int value;
  struct ssm_heap_object_t *object;
};
```

and the on-heap object:

```
struct ssm_heap_object_t {
 struct ssm_smm_md header;
 struct ssm_object_t payload[1];
};
```

This should fix the apparent cyclic type dependency because the C compiler doesn't need to know the size of the struct ssm_heap_object_t when it's allocating a pointer for it.

Although it's a little redundant, it probably makes sense to introduce a new struct type for each user-defined ADT, e.g.,

```
struct MyBool {
  struct ssm_mm_md header;
  struct ssm_object_t payload[2]; // max number of fields
};
```

It's redundant because every ADT with two fields max will look like this, but that's not terrible The advantage here is that we can generate code like `sizeof(struct MyBool)` and let the C compiler compute the right number of bytes.

Furthermore, we can "safely" generate code like myvar.object->payload[5] even though the object field is only an array of size 1 in the original type.